### ISSUE TITLE
Version 2.0

### ISSUE THREAD

**Issue by amcdnl** [Swimlane/angular-model-factory#56]
Version 2.0 will do:

- ES6 Classes with getter/setter/decorators
- Models without services

Idea API:

    export class Animal : ModelBase {

      @default
      color(){
        return 'white';
      },

      // i'm not even sure this is needed
      // anymore with es6 getters/setters
      @mapping
      food(){
        if(this.food === 1) return 'meat';
        if(this.food === 2) return 'bamboo';
      }

    }

    function AnimalService($model){
      return $model('animal', {
        actions: {
          // same as before
        },
        asClass: Animal
      })
    }

**Comment by juristr** [comment]
I'd propose the following upgrade steps while working towards v2.0 (or 1.0 :wink: )

- [ ] upgrade the existing codebase setup to one as used in the [angular-systemjs-seed proj](https://github.com/Swimlane/angular-systemjs-seed), that is with Gulp, ES6 etc. 
- [ ] create a branch `model-factory-es6` (or similar) within which we start developing the new version that takes advantage of new ES6 features. We could start building them against the existing test suite, in order to have a guideline of whether we cover the current features of v0.x (obviously tests need to be adjusted as we introduce breaking changes, but many should run without modifications I guess).
- [ ] once we have quite a stable version with the minimal usable feature set implemented, I'd suggest to immediately merge back into master and start releasing incrementally as we proceed (possibly totally automated as described #55 :smile: )

By first upgrading the tooling to Babel, SystemJS and ES6, we introduce the flexibility of being able to continue maintain the ES5 code, but still take advantage of the ES6 modules to more easily split out things into different files. Also migrating back to master from the "es6" branch would be easier afterwards.

Ok, that were my 2 cents. Going to sleep now, tired today :wink:

**Comment by Ye-Ting** [comment]
Wow!  It sounds very cool!

**Comment by amcdnl** [comment]
Found out a few new items based on the Angular2 Survey results yesterday: http://angularjs.blogspot.com/2015/09/angular-2-survey-results.html

*Angular 2 Data*
This library isn’t available yet (which is why it wasn’t in the survey), but is being modeled after Ember Data to provide a higher level interface for your data in Angular 2. Other data libraries may be used in conjunction at lower levels. We talked with the Angular core team about Angular 2 Data on Angular Air.

Given Angular2 will come with a more rich model library, I'm very interested to see what gaps it fills that our library handles.

One other thing, They mentioned a lot of people are using/planning to use RxJS ( This library implements observables, which are objects that model push-based data sources ).  They also mention this will be included in Angular2: 'this is why RxJS is used in the Angular 2 core http module'.  I tried to keep our library light as possible but if most of you are already using this or plan to, it would be VERY interesting to include this in some shape or fashion in v2 release.

Once some of the internals settle down w/ changes, I plan to have Angular2 support in v2 too.

**Comment by amcdnl** [comment]
Interesting undo/redo functionality we might want for 2.0 - https://github.com/omnidan/redux-undo

**Comment by amcdnl** [comment]
We should use generators like : http://mongorito.com/

**"Mongorito is based on ES6 generators, so you will not need to write a single callback (or promise). Enjoy the comfort of writing synchronous code, while retaining the power of Node.js' async nature."**

       yield model.save();
       model.doWhatever()

**Comment by amcdnl** [comment]
Would like to get feedback to whether people like the event listeners or would prefer 'hooks' on the static instance.  I think pub/sub on the rootsope is kinda frowned on in A1.x

**Comment by juristr** [comment]
@amcdnl Good idea. We could write it in the Gitter chatroom and/or open up an issue where we ask for it, referencing people where we know they use model-factory (from forks, stars, past issues...). I'll ask internally my devs what they think about it.

Do you have a concrete code sample (past -> proposed) to show. That'd help when asking.

**Comment by juristr** [comment]
> EDIT: Since A2 is getting rid of scope, we should do this instead

@amcdnl yep, I think that fits well, maybe even in a combination with RxJS (as I briefly experimented with [here](http://juristr.com/blog/2015/07/learning-ng-prepare-ng2/#UseObservablesinsteadof$watch)).

**Comment by NewMediaRoc** [comment]
Hey all, I was wondering how much work is left on version 2.0? If you're close but need some help, I'd be glad to jump in if I can.

**Comment by amcdnl** [comment]
@NewMediaRoc sadly been slammed w/ other projects.  very much still on my todo list, my aim is to make something that works w/ both a1 and a2 but not sure how feasible that is. maybe a similar API is a diff engine under the hood.

I started a branch, feel free to fork and go at it.  The big different is the use of classes and the `asClass` property. I also broke out all the bundled extras into mixins and broke it out to have a 'base' where it wasn't tied to a http service too. Hit me up on gitter if you wanna chat more!
